test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java

```diff
@@ -177,8 +179,23 @@ public boolean implies(ProtectionDomain domain, Permission permission) {
     private static void addClassCodebase(Map<String, URL> codebases, String name, String classname) {
         try {
             Class<?> clazz = BootstrapForTesting.class.getClassLoader().loadClass(classname);
-            if (codebases.put(name, clazz.getProtectionDomain().getCodeSource().getLocation()) != null) {
-                throw new IllegalStateException("Already added " + name + " codebase for testing");
+            URL location = clazz.getProtectionDomain().getCodeSource().getLocation();
+            try {
+                // if we find the class in a file (jar), make sure that jar is not already in the codebases map
+                // this will be the case when using test framework externally
+                Path filePath = PathUtils.get(location.toURI());
+                if (Files.isRegularFile(filePath)) {
+                    String fileName = filePath.getFileName().toString();
+                    if (!codebases.containsKey(name) && !codebases.containsKey(fileName)) {
```

Comment:
We prefer using `== false` instead of `!` as it the first is more obvious when scanning code.